src/builders.ts

```diff
@@ -1,18 +1,122 @@
 import { isAbsolute, relative } from 'path';
+import { EOL } from 'os';
 import slash = require('slash');
 import { createHash } from 'crypto';
 import {
   DaysToDecay,
   GenericLintData,
-  Location,
+  Operation,
   TodoConfig,
-  TodoData,
-  TodoDataV1,
   TodoDataV2,
   TodoDates,
   TodoFileFormat,
+  TodoFileHash,
+  TodoFilePathHash,
 } from './types';
 import { getDatePart } from './date-utils';
+import { todoDirFor } from './io';
+import TodoMatcher from './todo-matcher';
+
+const SEPARATOR = '|';
+
+export function buildFromTodoOperations(
+  todoOperations: string[]
+): Map<TodoFilePathHash, TodoMatcher> {
+  const existingTodos = new Map<TodoFilePathHash, TodoMatcher>();
+
+  for (const todoOperation of todoOperations) {
+    const [
+      operation,
+      engine,
+      ruleId,
+      line,
+      column,
+      endLine,
+      endColumn,
+      source,
+      fileFormat,
+      createdDate,
+      warnDate,
+      errorDate,
+      ...filePathSegments
+    ] = todoOperation.split(SEPARATOR);
+
+    // The only case where we need to join back on the separator is when the filePath itself
+    // contains a pipe ('|') char. Normal filePaths will simply join without the separator.
+    const filePath = filePathSegments.join(SEPARATOR);
+    const todoFileDir = todoDirFor(filePath);
+
+    if (!existingTodos.has(todoFileDir)) {
+      existingTodos.set(todoFileDir, new TodoMatcher());
+    }
+
+    const matcher = existingTodos.get(todoFileDir);
+
+    matcher?.addOrRemove(<Operation>operation, {
+      engine,
+      ruleId,
+      filePath,
+      fileFormat: Number.parseInt(fileFormat, 10),
+      range: {
+        start: {
+          line: Number.parseInt(line, 10),
+          column: Number.parseInt(column, 10),
+        },
+        end: {
+          line: Number.parseInt(endLine, 10),
+          column: Number.parseInt(endColumn, 10),
+        },
+      },
+      source,
+      createdDate: Number.parseInt(createdDate, 10),
+      warnDate: Number.parseInt(warnDate, 10),
+      errorDate: Number.parseInt(errorDate, 10),
```

Comment:
It is (and should continue to be) possible to have no warnDate / errorDate (it's just uncommon since you have to go out of your way to avoid our defaulting), I think in those cases this `parseInt` will result in `NaN`

Reply:
I've fixed this in a later PR, so going to resolve this one!